File: sdk/python/src/vorpal_sdk/artifact/uv.py
"""uv toolchain artifact (Astral standalone release).

Mirrors ``sdk/typescript/src/artifact/uv.js``.

HASH-ENFORCEMENT (C3 foundation): on ``uv sync --frozen``, uv verifies each package
against the per-package hashes carried in ``uv.lock`` and rejects any content-hash
mismatch. That hashed-lock verification IS the require-hashes enforcement surface the
build helpers wire — there is no ``uv sync --require-hashes`` CLI flag.

PROVENANCE — no inline digest (ADR 0001 Part A); see ``cpython.py`` for the rationale.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from vorpal_sdk.api.artifact import artifact_pb2
from vorpal_sdk.artifact import Artifact, ArtifactSource
from vorpal_sdk.artifact.cpython import cpython_target
from vorpal_sdk.step import shell

if TYPE_CHECKING:
    from vorpal_sdk.context import ConfigContext

# Canonical conforming copy of the build-target uv toolchain pin (Astral release).
# Byte-equal to the Rust canonical (ADR 0001 Part B).
DEFAULT_UV_VERSION = "0.10.11"


class Uv:
    """Builder for the uv toolchain artifact."""

    def __init__(self) -> None:
        self._version = DEFAULT_UV_VERSION

    def with_version(self, version: str) -> Uv:
        self._version = version
        return self

    def build(self, context: ConfigContext) -> str:
        name = "uv"
        system = context.get_system()

        source_target = cpython_target(system)
        source_version = self._version
        source_path = (
            f"https://sdk.vorpal.build/source/"
            f"uv-{source_version}-{source_target}.tar.gz"
        )

        source = ArtifactSource(name, source_path).build()

        # Astral standalone release unpacks to uv-{triple}/uv at the tarball root.
        step_script = f"""mkdir -p "$VORPAL_OUTPUT/bin"
cp -p "./source/{name}/uv-{source_target}/uv" "$VORPAL_OUTPUT/bin/uv"
chmod +x "$VORPAL_OUTPUT/bin/uv"
"""
        steps = [shell(context, [], [], step_script, [])]
        systems = [
            artifact_pb2.AARCH64_DARWIN,
            artifact_pb2.AARCH64_LINUX,
            artifact_pb2.X8664_DARWIN,
            artifact_pb2.X8664_LINUX,
        ]

        return (
            Artifact(name, steps, systems)
            .with_aliases([f"{name}:{source_version}"])
            .with_sources([source])
            .build(context)
        )

File: sdk/go/pkg/artifact/uv.go
package artifact

import (
	"fmt"

	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

const defaultUvVersion = "0.10.11"

// Uv builds the Astral uv toolchain (standalone release).
//
// HASH-ENFORCEMENT (C3 foundation): on "uv sync --frozen", uv-0.10.11 verifies each
// package against the per-package hashes in uv.lock and rejects any content-hash
// mismatch. There is no "uv sync --require-hashes" CLI flag (--require-hashes is uv's
// pip-interface flag); the hashed-lock verification IS the enforcement surface.
//
// PROVENANCE — no inline WithDigest (ADR 0001 Part A). The canonical pin is the per-triple
// Vorpal.lock entry captured via --unlock; until then the HTTP source is intentionally
// unpinned and the C1 mint gate fails the build closed. See Cpython for the full
// two-link provenance rationale.
func Uv(context *config.ConfigContext) (*string, error) {
	name := "uv"

	system := context.GetTarget()

	sourceTarget, err := CpythonTarget(system)
	if err != nil {
		return nil, err
	}

	sourceVersion := defaultUvVersion
	sourcePath := fmt.Sprintf(
		"https://sdk.vorpal.build/source/uv-%s-%s.tar.gz",
		sourceVersion, sourceTarget,
	)

	source := NewArtifactSource(name, sourcePath).Build()

	// Astral standalone release unpacks to uv-{triple}/uv at the tarball root.
	stepScript := fmt.Sprintf(`mkdir -p "$VORPAL_OUTPUT/bin"
cp -p "./source/%s/uv-%s/uv" "$VORPAL_OUTPUT/bin/uv"
chmod +x "$VORPAL_OUTPUT/bin/uv"
`, name, sourceTarget)

	step, err := Shell(context, []*string{}, []string{}, stepScript, nil)
	if err != nil {
		return nil, err
	}

	systems := []api.ArtifactSystem{
		api.ArtifactSystem_AARCH64_DARWIN,
		api.ArtifactSystem_AARCH64_LINUX,
		api.ArtifactSystem_X8664_DARWIN,
		api.ArtifactSystem_X8664_LINUX,
	}

	return NewArtifact(name, []*api.ArtifactStep{step}, systems).
		WithAliases([]string{fmt.Sprintf("%s:%s", name, sourceVersion)}).
		WithSources([]*api.ArtifactSource{&source}).
		Build(context)
}
